homework_solution.md

```diff
@@ -0,0 +1,474 @@
+
+      ```
+      SELECT setval(pg_get_serial_sequence('stations', 'id'), (SELECT MAX(id) FROM stations));
+      ```
+
+   6. Finally, add entries for records where a name exists, but the id is null, relying on the seqquence to generate an id. This has to be done in 2 steps, so we can exclude the records from the first insertions when we run the second one, avoiding duplicates.
+
+      ```
+      INSERT INTO stations (
+         SELECT nextval(pg_get_serial_sequence('stations', 'id')), from_station_name
+         FROM trips
+         WHERE
+            from_station_id IS NULL
+         AND
+            from_station_name NOT IN (SELECT name from stations)
+         GROUP BY from_station_name
+      );
+      INSERT INTO stations(
+         SELECT nextval(pg_get_serial_sequence('stations', 'id')), to_station_name
+         FROM trips
+         WHERE
+            to_station_id IS NULL
+         AND
+            to_station_name NOT IN (SELECT name from stations)
+         GROUP BY to_station_name
+      );
+      ```
+
+2. _**Should we add any indexes to the stations table, why or why not?**_
+
+   Not right now. We already have an index on `id`, because it's the table's pkey. For our current usage, that should suffice.
+
+3. _**Fill in the missing data in the `trips` table based on the work you did above**_
+
+   ### Steps
+
+   1. We _could_ update the names in `trips`, like so:
+
+      ```
+      UPDATE trips t
+      SET
+         from_station_name = (SELECT name FROM stations s WHERE t.from_station_id = s.id),
+         to_station_name = (SELECT name FROM stations s WHERE t.from_station_id = s.id)
+      ```
+
+   2. However, it's best to just delete the station names column, so the `stations` table becomes the single source of truth for that (but we won't do that just yet!!)
+
+      ```
+      ALTER TABLE trips_backup
+      DROP COLUMN from_station_name,
+      DROP COLUMN to_station_name;
+      ```
+
+   3. Before dropping the columns, we have to populate the station ids columns where they're null
+
+      ```
+      UPDATE trips t
+      SET from_station_id = (SELECT id FROM stations s WHERE t.from_station_name = s.name),
+      to_station_id  = (SELECT id FROM stations s WHERE t.to_station_name = s.name)
+      WHERE from_station_id IS NULL OR to_station_id IS NULL
+      ```
+
+Now all of our `trips` should have stations ids that can be joined on the `stations` table when we need to get more information.
+
+# Part 2: Missing Date Data
+
+1.  _**What's the inconsistency in date formats? You can assume that each quarter's trips are numbered sequentially, starting with the first day of the first month of that quarter.**_
+
+    There are many. We can examine them by running (this uses the fairly safe assumption of consistency within each of the original csv, so looking at the first record for a given file provides insight into the whole file, and looking into the start date format gives insight into the end date format).
```

Comment:
[question]
Why is the assumption that the date formats are consistent a safe one?

Reply:
I'm not sure if safe is the right one, but it's probably a fair one, given the fact that each group of dates came in as a set, identified by the file.
Some inconsistencies would be caught as errors if they existed (for instance, if we assumed `25/12`, and one record in that set had `12/25`, we'd get an error indicating an invalid month, and would need to revisit that assumption. Across so many records, that'd be likely to happen if there were inconsistencies.
Just in case though, the original date strings aren't removed, to allow for further verification and sanitization if needed.

Reply:
Well put! It's important to sanity-check these things from a practical perspective. It's easy to think you've covered all the cases, when you've missed one or two.